File: codes/matlab/ch20/sinc_two_masters.m
function sinc_two_masters()
%% sinc_two_masters - Chapter 20, Etude 20.3.  Sinc expansions: two masters, span and spacing.
% Author: Dr. Denys Dutykh.

    script_dir = fileparts(mfilename('fullpath'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    N_fix = 48; hs = linspace(0.15, 1.8, 80);
    err_sweep = arrayfun(@(h) sinc_err(N_fix, h), hs);

    Ns = [8 12 16 24 32 48 64 96 128 192];
    h_opts = sqrt(pi^2 ./ (2*Ns));
    err_opt = arrayfun(@(i) sinc_err(Ns(i), h_opts(i)), 1:length(Ns));

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    fig = figure('Position',[100 100 1100 800],'Color','w');
    [~, jstar] = min(err_sweep);
    h_theory = sqrt(pi^2/(2*N_fix));

    % ---- (a) V-shape: empirical error vs h ---------------------------
    subplot(2,2,1);
    semilogy(hs, err_sweep, 'Color', CORAL, 'LineWidth', 1.2); hold on;
    xline(hs(jstar), ':', 'Color', NAVY, 'LineWidth', 1.0, ...
          'DisplayName', sprintf('h* = %.2f empirical', hs(jstar)));
    xline(h_theory, '--', 'Color', TEAL, 'LineWidth', 1.0, ...
          'DisplayName', sprintf('sqrt(pi^2/(2N)) = %.2f', h_theory));
    xlabel('h'); ylabel('max error'); grid on; box on;
    title(sprintf('(a) Fix N=%d, vary h', N_fix));
    legend('show', 'Location', 'northwest');

    % ---- (b) NEW: two-master decomposition ---------------------------
    subplot(2,2,2);
    hs_an = linspace(min(hs), max(hs), 400);
    E_W  = (4/pi) .* exp(-pi^2 ./ (2*hs_an));
    E_DT = 4 .* exp(-N_fix .* hs_an ./ 2);
    semilogy(hs_an, E_W,  '--', 'Color', NAVY,  'LineWidth', 1.0, ...
             'DisplayName', 'E_W (bandwidth) ~ exp(-pi^2/(2h))'); hold on;
    semilogy(hs_an, E_DT, '--', 'Color', CORAL, 'LineWidth', 1.0, ...
             'DisplayName', 'E_{DT} (grid span) ~ exp(-Nh/2)');
    semilogy(hs_an, E_W + E_DT, 'Color', TEAL, 'LineWidth', 1.4, ...
             'DisplayName', 'sum: E_W + E_{DT}');
    semilogy(hs, err_sweep, 'Color', [0.5 0.5 0.5], 'LineWidth', 0.7, ...
             'DisplayName', 'empirical V');
    xline(hs(jstar), ':', 'Color', NAVY, 'LineWidth', 0.8, 'HandleVisibility', 'off');
    xlabel('h'); ylabel('error'); grid on; box on;
    title('(b) Two masters: bandwidth vs grid-span');
    legend('show', 'Location', 'southwest', 'FontSize', 7);
    ylim([1e-12 1e1]);

    % ---- (c) Subgeometric convergence at h_opt -----------------------
    subplot(2,2,3);
    semilogy(sqrt(Ns), err_opt, '-o', 'Color', TEAL, ...
             'MarkerFaceColor', 'w', 'LineWidth', 1.1, ...
             'DisplayName', 'sinc at h = sqrt(pi^2/(2N))'); hold on;
    semilogy(sqrt(Ns), exp(-pi*sqrt(Ns/2)), ':', 'Color', NAVY, 'LineWidth', 1.0, ...
             'DisplayName', 'exp(-pi sqrt(N/2))');
    xlabel('sqrt(N)'); ylabel('max error'); grid on; box on;
    title('(c) Subgeometric convergence at h = h*');
    legend('show', 'Location', 'northeast');

    % ---- (d) Grid cartoon -------------------------------------------
    subplot(2,2,4);
    hold on;
    for k = 1:2
        Nc = [8 32]; offs = [0.7 0.3]; cs = {CORAL, TEAL};
        hk = sqrt(pi^2/(2*Nc(k)));
        j = -floor(Nc(k)/2):floor(Nc(k)/2);
        yj = j * hk;
        scatter(yj, offs(k)*ones(size(yj)), 16, cs{k}, 'x', ...
                'DisplayName', sprintf('N=%d, h=%.2f', Nc(k), hk));
    end
    y_plot = linspace(-10, 10, 401);
    plot(y_plot, 0.05 + 0.6./cosh(y_plot), 'Color', NAVY, 'LineWidth', 1.0, ...
         'HandleVisibility', 'off');
    xlim([-10 10]); ylim([0 1.2]); box on;
    xlabel('y'); title('(d) Span and spacing grow together');
    legend('show', 'Location', 'northeast');

    set(fig, 'PaperPositionMode', 'auto');
    print(fig, fullfile(out_dir, 'sinc_two_masters.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'sinc_two_masters.png'), '-dpng');
    fprintf('[20.3-matlab] figure saved\n');
end

function e = sinc_err(N, h)
    j = -floor(N/2):floor(N/2);
    yj = j * h;
    fj = 1 ./ cosh(yj);
    y = linspace(-20, 20, 4001);
    z = (y(:) - yj) / h;
    S = sin(pi*z) ./ (pi*z);
    S(abs(z) < 1e-14) = 1;
    approx = S * fj(:);
    e = max(abs(approx(:) - 1./cosh(y(:))));
end

File: codes/julia/ch20/sinc_two_masters.jl
# sinc_two_masters.jl
# Chapter 20: Spectral Methods on Unbounded Intervals
# Computational Etude 20.3: Sinc -- two masters, span and spacing.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using Printf

target(y) = 1.0 / cosh(y)

function sinc_approx(y, N, h)
    j = (-div(N, 2)):div(N, 2)
    yj = j .* h
    fj = target.(yj)
    vals = zeros(length(y))
    for (i, yi) in enumerate(y)
        acc = 0.0
        for k in 1:length(yj)
            z = (yi - yj[k]) / h
            s = abs(z) < 1e-14 ? 1.0 : sin(pi * z) / (pi * z)
            acc += s * fj[k]
        end
        vals[i] = acc
    end
    return vals
end

function sinc_error(N, h)
    y = collect(range(-20.0, 20.0, length=4001))
    approx = sinc_approx(y, N, h)
    return maximum(abs.(approx .- target.(y)))
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch20", "julia")
    mkpath(outdir)

    N_fix = 48
    hs = collect(range(0.15, 1.8, length=80))
    err_sweep = [sinc_error(N_fix, h) for h in hs]

    Ns = [8, 12, 16, 24, 32, 48, 64, 96, 128, 192]
    h_opts = [sqrt(pi^2 / (2 * N)) for N in Ns]
    err_opt = [sinc_error(N, h) for (N, h) in zip(Ns, h_opts)]

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    fig = Figure(size=(1100, 800))

    j_star = argmin(err_sweep)
    h_theory = sqrt(pi^2 / (2 * N_fix))

    # ---- (a) V-shape: empirical error vs h ----
    ax1 = Axis(fig[1, 1], xlabel="h", ylabel="max error",
               yscale=log10, title="(a) Fix N=$N_fix, vary h")
    lines!(ax1, hs, err_sweep; color=CORAL, linewidth=1.2)
    vlines!(ax1, [hs[j_star]]; linestyle=:dot, color=NAVY,
            label="h* = $(round(hs[j_star]; digits=2)) empirical")
    vlines!(ax1, [h_theory]; linestyle=:dash, color=TEAL,
            label="sqrt(pi^2/(2N)) = $(round(h_theory; digits=2))")
    axislegend(ax1; position=:lt, labelsize=9)

    # ---- (b) NEW: two-master decomposition ----
    hs_an = collect(range(minimum(hs), maximum(hs), length=400))
    E_W  = (4 / pi) .* exp.(-pi^2 ./ (2 .* hs_an))
    E_DT = 4 .* exp.(-N_fix .* hs_an ./ 2)
    ax2 = Axis(fig[1, 2], xlabel="h", ylabel="error",
               yscale=log10,
               title="(b) Two masters: bandwidth vs grid-span",
               limits=(nothing, (1e-12, 1e1)))
    lines!(ax2, hs_an, E_W;  color=NAVY,  linestyle=:dash, linewidth=1.0,
           label="E_W (bandwidth) ~ exp(-pi^2/(2h))")
    lines!(ax2, hs_an, E_DT; color=CORAL, linestyle=:dash, linewidth=1.0,
           label="E_DT (grid span) ~ exp(-Nh/2)")
    lines!(ax2, hs_an, E_W .+ E_DT; color=TEAL, linewidth=1.4,
           label="sum: E_W + E_DT")
    lines!(ax2, hs, err_sweep; color=:gray, linewidth=0.7,
           label="empirical V")
    vlines!(ax2, [hs[j_star]]; linestyle=:dot, color=NAVY, linewidth=0.8)
    axislegend(ax2; position=:lb, labelsize=8)

    # ---- (c) Subgeometric convergence at h_opt ----
    ax3 = Axis(fig[2, 1], xlabel="sqrt(N)", ylabel="max error",
               yscale=log10, title="(c) Subgeometric convergence at h = h*")
    scatterlines!(ax3, sqrt.(Ns), err_opt; color=TEAL,
                  markercolor=:white, strokecolor=TEAL, strokewidth=1.0,
                  label="sinc at h = sqrt(pi^2/(2N))")
    lines!(ax3, sqrt.(Ns), exp.(-pi .* sqrt.(Ns ./ 2));
           color=NAVY, linestyle=:dot, label="exp(-pi sqrt(N/2))")
    axislegend(ax3; position=:rt, labelsize=9)

    # ---- (d) Grid cartoon ----
    ax4 = Axis(fig[2, 2], xlabel="y",
               title="(d) Span and spacing grow together",
               limits=((-10, 10), (0, 1.2)))
    for (k, (Nc, offset)) in enumerate([(8, 0.7), (32, 0.3)])
        hk = sqrt(pi^2 / (2 * Nc))
        j = (-div(Nc, 2)):div(Nc, 2)
        yj = j .* hk
        scatter!(ax4, yj, fill(offset, length(yj));
                 color=(k == 1 ? CORAL : TEAL), marker=:xcross, markersize=10,
                 label="N=$Nc, h=$(round(hk; digits=2))")
    end
    y_plot = collect(range(-10.0, 10.0, length=401))
    lines!(ax4, y_plot, 0.05 .+ 0.6 .* target.(y_plot); color=NAVY, linewidth=1.0)
    axislegend(ax4; position=:rt, labelsize=9)

    save(joinpath(outdir, "sinc_two_masters.pdf"), fig)
    save(joinpath(outdir, "sinc_two_masters.png"), fig)
    @printf("[20.3-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end
